### PR TITLE
Add initial Sample-Layer Attention for GPTQ (PyTorch)

### PR DESCRIPTION
**Pull request by irenaby**
## Pull Request Description:
Add hessian estimation per image hash.
Add sample-layer attention distillation loss.
Add weights per layer to soft round loss.
Update GPTQ config and its generation for sample layer attention.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/hessian/activation_hessian_scores_calculator_pytorch.py [hunk 1]
```diff
@@ -55,98 +56,145 @@ def __init__(self,
                                                                        hessian_scores_request=hessian_scores_request,
                                                                        num_iterations_for_approximation=num_iterations_for_approximation)
 
+    def forward_pass(self):
```
Comment (irenaby) [inline model_compression_toolkit/core/pytorch/hessian/activation_hessian_scores_calculator_pytorch.py:59]: no change, just extracted to method

File: model_compression_toolkit/core/pytorch/hessian/activation_hessian_scores_calculator_pytorch.py [hunk 2]
```diff
@@ -55,98 +56,145 @@ def __init__(self,
...
+        else:
+            raise NotImplementedError(f'{self.hessian_request.granularity} is not supported')
+
+        # Convert results to list of numpy arrays
+        hessian_results = [torch_tensor_to_numpy(h) for h in hessian_scores]
+        return hessian_results
+
+    def _compute_per_tensor(self, output, target_activation_tensors):
+        assert self.hessian_request.granularity == HessianScoresGranularity.PER_TENSOR
+        ipts_hessian_approx_scores = [torch.tensor([0.0], requires_grad=True, device=output.device)
```
Comment (irenaby) [inline model_compression_toolkit/core/pytorch/hessian/activation_hessian_scores_calculator_pytorch.py:140]: no change (except line 177), just extracted to method

File: model_compression_toolkit/core/common/hessian/hessian_scores_request.py [hunk 3]
```diff
@@ -40,6 +40,14 @@ class HessianScoresGranularity(Enum):
     PER_TENSOR = 2
 
 
+class HessianEstimationDistribution(str, Enum):
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/hessian_scores_request.py:43]: why do we need this? I think you can use the rademacher in all cases and remove this enum and the addition to the gptq config.
The results (on average) should be similar for both methods if Im not mistaken (unless you seen a different behavior?)

File: model_compression_toolkit/gptq/pytorch/gptq_training.py [hunk 4]
```diff
@@ -276,6 +289,40 @@ def micro_training_loop(self,
...
+
+        if len(input_tensors) > 1:
+            raise NotImplementedError('Sample-Layer attention is not currently supported for networks with multiple inputs')
+
+        scores = []
+        batch = input_tensors[0]
+        img_hashes = [self.hessian_service.calc_image_hash(img) for img in batch]
+        for img_hash in img_hashes:
+            if img_hash not in self.hessian_score_per_image_per_layer:
+                score_per_image_layer_per = self._compute_sample_layer_attention_scores(input_tensors)
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/pytorch/gptq_training.py:316]: I guess you meant "score_per_image_per_layer"?
In general, I think this function needs to be rewritten cause it's hard to track what's going on here.

File: model_compression_toolkit/gptq/pytorch/quantization_facade.py [hunk 5]
```diff
@@ -105,19 +108,37 @@ def get_pytorch_gptq_config(n_epochs: int,
...
+
         if isinstance(gradual_activation_quantization, bool):
             gradual_quant_config = GradualActivationQuantizationConfig() if gradual_activation_quantization else None
         elif isinstance(gradual_activation_quantization, GradualActivationQuantizationConfig):
             gradual_quant_config = gradual_activation_quantization
-        else:
+        else:    # pragma: no cover
             raise TypeError(f'gradual_activation_quantization argument should be bool or '
-                            f'GradualActivationQuantizationConfig, received {type(gradual_activation_quantization)}')    # pragma: no cover
+                            f'GradualActivationQuantizationConfig, received {type(gradual_activation_quantization)}')
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/pytorch/quantization_facade.py:135]: Please, add a TODO to replace the 'no cover' since this case should be tested.

File: model_compression_toolkit/gptq/pytorch/quantizer/soft_rounding/soft_quantizer_reg.py [hunk 6]
```diff
@@ -40,32 +40,36 @@ def __init__(self, beta_scheduler: Callable[[int], float]):
 
         self.count_iter = 0
 
-    def __call__(self, model: nn.Module, entropy_reg: float):
+    def __call__(self, model: nn.Module, entropy_reg: float, layer_weights: torch.Tensor = None):
```
Comment (reuvenperetz) [inline model_compression_toolkit/gptq/pytorch/quantizer/soft_rounding/soft_quantizer_reg.py:43]: If I'm not mistaken, the default case (where the weighting is all ones) is taking care outside of this function. If this is the case, maybe it is may be better to remove the default value of layer_weights?

File: model_compression_toolkit/core/common/hessian/hessian_scores_request.py (hunk 3 above)
Comment (irenaby) [reply model_compression_toolkit/core/common/hessian/hessian_scores_request.py:43]: I think I saw different results, not sure. In any case it changes the existing behavior so I don't think this belongs to this PR. This isn't exposed to user anyway, so there is no problem to remove it later.

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 7]
```diff
@@ -143,7 +144,11 @@ def compute_hessian_based_weights(self) -> np.ndarray:
             return np.asarray([1 / num_nodes for _ in range(num_nodes)])
 
         # Fetch hessian approximations for each target node
-        compare_point_to_hessian_approx_scores = self._fetch_hessian_approximations()
+        # TODO this smells like a potential bug. In hessian calculation target nodes are topo sorted and results are returned
+        # TODO also target nodes are replaced for reuse. Does this work correctly?
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/gptq/common/gptq_training.py:148]: target_nodes are sorted in place in the request during processing, so the order that the caller will receive might be different from the order he defined in the request. It's also bad practice to modify the request, but that's another story.

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 8]
```diff
@@ -172,29 +177,55 @@ def compute_hessian_based_weights(self) -> np.ndarray:
...
+        hessian_score_per_image_per_layer = {}
+        # If hessian batch is smaller than inputs batch, split it to hessian batches. If hessian batch is larger,
+        # it's currently ignored (TODO)
+        for i in range(0, inputs_batch[0].shape[0], hessian_batch_size):
+            inputs = [t[i: i+hessian_batch_size] for t in inputs_batch]
+            hessian_score_per_image_per_layer.update(
+                self.hessian_service.compute_trackable_per_sample_hessian(request, inputs)
+            )
+        for img_hash, v in hessian_score_per_image_per_layer.items():
+            hessian_score_per_image_per_layer[img_hash] = {k: t.max(axis=0) for k, t in v.items()}
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/gptq/common/gptq_training.py:203]: That's the definition of the sample-layer attention. My understanding is that we are trying to approximate the upper bound, so we take the max among channels (this is per image so axis=0 is channels)

File: model_compression_toolkit/gptq/pytorch/gptq_training.py [hunk 9]
```diff
@@ -105,8 +105,17 @@ def _get_total_grad_steps():
...
                                                                   trainable_bias,
                                                                   trainable_threshold)
-
-        self.weights_for_average_loss = to_torch_tensor(self.compute_hessian_based_weights())
+        hessian_cfg = self.gptq_config.hessian_weights_config
+        self.use_sample_layer_attention = hessian_cfg.per_sample
+        self.hessian_score_per_layer = None    # for fixed layer weights
+        self.hessian_score_per_image_per_layer = None    # for sample-layer attention
+        if self.use_sample_layer_attention:
+            assert (hessian_cfg.norm_scores is False and hessian_cfg.log_norm is False and
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/gptq_training.py:113]: Changed to NotImplementedError and added comment.

File: model_compression_toolkit/gptq/pytorch/gptq_training.py (hunk 4 above)
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/gptq_training.py:316]: Everything should be rewritten. Added more comments, hope it's clearer now.

File: model_compression_toolkit/gptq/pytorch/quantizer/soft_rounding/soft_quantizer_reg.py (hunk 6 above)
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/quantizer/soft_rounding/soft_quantizer_reg.py:43]: You are not mistaken. I will remove it for now due to missing testing/coverage, but in principle I don't see a connection between the two. I think in this case it makes sense to have a default behavior, the fact that it was eventually more convenient to pass explicit weights in the specific use case should not really affect this.

File: model_compression_toolkit/gptq/pytorch/quantization_facade.py (hunk 5 above)
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/quantization_facade.py:135]: I think that's true for all no cover throughout the code

File: model_compression_toolkit/core/pytorch/hessian/activation_hessian_scores_calculator_pytorch.py [hunk 10]
```diff
@@ -55,98 +56,145 @@ def __init__(self,
...
+                                       inputs=ipt_tensor,
+                                       retain_graph=True)[0]
+                hessian_approx_scores = hess_v ** 2
+                rank = len(hess_v.shape)
+                if rank > 2:
+                    hessian_approx_scores = torch.mean(hessian_approx_scores, dim=tuple(range(2, rank)))
+
+                # Update node Hessian approximation mean over random iterations
+                ipts_hessian_approx_scores[i] = (j * ipts_hessian_approx_scores[i] + hessian_approx_scores) / (j + 1)
+
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/pytorch/hessian/activation_hessian_scores_calculator_pytorch.py:199]: It's not exactly the same. I'm sure it can be rewritten, but it's not as straightforward as in other places that were extracted, so was not a priority.

**Review by reuvenperetz — APPROVED** [review]
